### PR TITLE
Speed limit

### PR DESCRIPTION
**Pull request by isaacsimmons**

### CODE REVIEW

File: src/Lhm/Chunker.php [hunk 1]
```diff
@@ -86,13 +90,25 @@ public function __construct(AdapterInterface $adapter, \Phinx\Db\Table $origin,
...
+            'stride' => 2000,
+            'strideMax' => 20000,
+            'targetQuerySeconds' => 0.75,
+        ];
 
         $this->primaryKey = $this->adapter->quoteColumnName($this->sqlHelper->extractPrimaryKey($this->origin));
 
+        $adapterFactory = AdapterFactory::instance();
+        $res = $this->adapter->query('SELECT HOST FROM INFORMATION_SCHEMA.PROCESSLIST WHERE COMMAND = \'Binlog Dump\'')->fetchAll();
+        foreach ($res as $row) {
```
Comment (isaacsimmons) [inline src/Lhm/Chunker.php]: do I need to pull $row['HOST'] instead?? I saw some snippets that make it seem that way and others where it doesn't

Comment (mc0) [reply src/Lhm/Chunker.php]: They both work (both is the default)

File: src/Lhm/Chunker.php [hunk 2]
```diff
@@ -86,13 +90,25 @@ public function __construct(AdapterInterface $adapter, \Phinx\Db\Table $origin,
...
-            ];
+            'stride' => 2000,
+            'strideMax' => 20000,
+            'targetQuerySeconds' => 0.75,
+        ];
 
         $this->primaryKey = $this->adapter->quoteColumnName($this->sqlHelper->extractPrimaryKey($this->origin));
 
+        $adapterFactory = AdapterFactory::instance();
+        $res = $this->adapter->query('SELECT HOST FROM INFORMATION_SCHEMA.PROCESSLIST WHERE COMMAND = \'Binlog Dump\'')->fetchAll();
```
Comment (mc0) [inline src/Lhm/Chunker.php]: Allow overriding this via a config in addition to this fetchAll

Comment (isaacsimmons) [reply src/Lhm/Chunker.php]: oh nuts.. forgot about that bit

File: src/Lhm/Chunker.php [hunk 3]
```diff
@@ -122,7 +152,28 @@ protected function execute()
...
+                && !empty($this->slaveAdapters)
+            ) {
+                foreach ($this->slaveAdapters as $slaveAdapter) {
+                    $result = $slaveAdapter->query('SHOW SLAVE STATUS')->fetchAll();
+                    foreach ($result as $row) {
+                        $lag = $result['Seconds_Behind_Master'];
+                        if ($lag > self::ALLOWED_SLAVE_LAG) {
+                            $this->logger->warning("Slave lag at $lag. Over allowed max, slowing down");
+                            sleep(10);
+                            $stride /= 2;
```
Comment (mc0) [inline src/Lhm/Chunker.php]: So I was wrong and said to change the stride, that's not what LHM/pt-osc are doing.

They ramp up or lower the sleep time between queries (e.g. sleep a minimum of 1ms, a maximum of 10s) depending on if the slaves are behind.

So something like:
```
$sleepTime = 1000;
if (overSlaveLag) {
    $sleepTime = min($sleepTime * 2, 1000000);
} else {
    $sleepTime = max($sleepTime / 2, 1000);
}
usleep($sleepTime);
```

**Review by mc0 — CHANGES_REQUESTED** [review]

**Review by mc0 — APPROVED** [review]

**Pull request merged**